Update NxSubmitMask design RSC-147

glenhunter opened the pull request:
https://issues.sonatype.org/browse/RSC-147

<img width="301" alt="Screen Shot 2020-10-23 at 2 42 43 PM" src="https://user-images.githubusercontent.com/1991443/97041787-07537e00-153e-11eb-9fe3-81045745a3ea.png">

rpokorny commented:
I've been lately second-guessing my original decision to include that goofy mid-line ellipsis in the example.  The point was just to demonstrate that you can change the punctuation, but everyone's first instinct when they see it is that it's a bug.  This would be a good time to change it to something else.

glenhunter commented:
> I've been lately second-guessing my original decision to include that goofy mid-line ellipsis in the example. The point was just to demonstrate that you can change the punctuation, but everyone's first instinct when they see it is that it's a bug. This would be a good time to change it to something else.

The goofy ellipsis is no more: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/4872c1af58de09e8cbbe621101b432ed7f625531

rpokorny commented on gallery/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@sonatype/react-shared-components-gallery",
-  "version": "2.5.0",
+  "version": "2.6.0",
```
Shouldn't this be a patch version?

glenhunter replied:
> Shouldn't this be a patch version?

Yeah it should: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/aac0a6d958e152b02a2c6080a060e75e9cb52bd2

rpokorny replied:
> Yeah it should: aac0a6d

Not .1?

glenhunter replied:
> > Yeah it should: [aac0a6d](https://github.com/sonatype/sonatype-react-shared-components/commit/aac0a6d958e152b02a2c6080a060e75e9cb52bd2)
> 
> Not .1?

You suggested patch instead of minor version so that means I back up from 2.6.0 to 2.5.x, then increment up 0.0.1 from there

rpokorny replied:
> You suggested patch instead of minor version so that means I back up from 2.6.0 to 2.5.x, then increment up 0.0.1 from there

But you went up by 0.0.2:
![image](https://user-images.githubusercontent.com/3630091/97352812-7fd17c00-1869-11eb-89a7-bed8f7536a2c.png)

glenhunter replied:
> > You suggested patch instead of minor version so that means I back up from 2.6.0 to 2.5.x, then increment up 0.0.1 from there
> 
> But you went up by 0.0.2:

Hmmm, odd, merged from Master and bumped again: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/3e9ac67844b18cbbbb5b4912d04cec8e5220863f

rpokorny commented on lib/src/components/NxSubmitMask/NxSubmitMask.scss:
```diff
@@ -6,11 +6,12 @@
   align-items: center;
 
-  background-color: rgba($nx-off-white, 0.9);
+  background-color: rgba(#555, 0.75);
```
I don't think this is technically valid css syntax, but scss seems to be converting it so I guess it's fine

glenhunter replied:
> I don't think this is technically valid css syntax, but scss seems to be converting it so I guess it's fine

Fixed here: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/12f0e8dbc13f4e55843306be3219208c4a9eaba3

glenhunter replied:
> > I don't think this is technically valid css syntax, but scss seems to be converting it so I guess it's fine
> 
> Fixed here: [12f0e8d](https://github.com/sonatype/sonatype-react-shared-components/commit/12f0e8dbc13f4e55843306be3219208c4a9eaba3)

This is a better fix, more consistent with how we do it elsewhere: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/eebce7f8a2e4002495a1b1d2ece4669d033a5c15

rpokorny commented on lib/src/components/NxSubmitMask/NxSubmitMask.scss:
```diff
@@ -30,17 +31,18 @@
   margin: 0;
-  padding: 22px 12px;
+  padding: $nx-spacing-l;
+  min-width: 124px;
```
This was intended as a border-box width but it's currently the content-box width

glenhunter replied:
> This was intended as a border-box width but it's currently the content-box width

Fixed here by changing it to `border-box`: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/12f0e8dbc13f4e55843306be3219208c4a9eaba3

rpokorny commented:
Should the content be centered when it is less than the minimum width?  It isn't currently.

rpokorny commented on lib/src/components/NxSubmitMask/NxSubmitMask.scss:
```diff
@@ -30,17 +31,18 @@
   .nx-submit-mask__message {
-    border-color: $nx-dark-green;
-    color: $nx-dark-green;
+    color: #099843;
```
I'm seeing #1a832b in zeplin

glenhunter replied:
> I'm seeing #1a832b in zeplin

I remember grabbing the green from Zeplin, but it's clearly wrong so it's fixed here: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/a501adad683c58369ef36082dd474034170848da

rpokorny commented:
Success icon has left margin; it shouldn't
![image](https://user-images.githubusercontent.com/3630091/97229337-98309080-17ae-11eb-8f6d-09a5e940f9a4.png)

rpokorny commented:
The Placement/Layout section of the mockups indicates that the message box should be 200px from the top of the viewport.  Fullscreen submit message boxes are currently centered in the viewport.  I'm also wondering if @dturner32 is aware that the masks are not fullscreen by default, and whether he intends that we continue to support that or that we drop it.

rpokorny commented on lib/src/components/NxSubmitMask/NxSubmitMask.tsx:
```diff
@@ -28,9 +30,15 @@ const NxSubmitMask: FunctionComponent<Props> =
+          { success ?
+            <>
+              <NxFontAwesomeIcon icon={faCheckCircle} className="nx-loading-spinner__icon" />
+              <span>{spinnerMessage || 'Loading…'}</span>
```
This `|| 'Loading...'` is wrong.  This is the success message here.  At the same time, `spinnerMessage` will always have a truthy value here based on the way it's created above.

glenhunter replied:
> This `|| 'Loading...'` is wrong. This is the success message here. At the same time, `spinnerMessage` will always have a truthy value here based on the way it's created above.

Fixed here: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/7eaad8766f8001a37f8fff6f890d10ea7a8be8af

rpokorny commented on lib/src/components/NxSubmitMask/NxSubmitMask.tsx:
```diff
@@ -28,9 +30,15 @@ const NxSubmitMask: FunctionComponent<Props> =
+              <NxFontAwesomeIcon icon={faCheckCircle} className="nx-loading-spinner__icon" />
+              <span>{spinnerMessage || 'Loading…'}</span>
+            </> :
+            <NxLoadingSpinner>{spinnerMessage}</NxLoadingSpinner>
```
I think the variables are a little bit tangled up.  What you have works but it uses variables names that aren't quite accurate to what the variables do.  What I'd do instead is pass the `message` prop directly to the `<NxLoadingSpinner>`, and then instead of having a variable called `spinnerMessage`, use the `successMessage` prop in the JSX construction for the success case, with the `|| 'Success!'` part in there for the default.

glenhunter replied:
> I think the variables are a little bit tangled up. What you have works but it uses variables names that aren't quite accurate to what the variables do. What I'd do instead is pass the `message` prop directly to the `<NxLoadingSpinner>`, and then instead of having a variable called `spinnerMessage`, use the `successMessage` prop in the JSX construction for the success case, with the `|| 'Success!'` part in there for the default.

I did a bunch of work on it here, I think it's much better now: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/7eaad8766f8001a37f8fff6f890d10ea7a8be8af

rpokorny commented on lib/src/components/NxSubmitMask/__tests__/NxSubmitMask.test.tsx:
```diff
@@ -42,17 +41,18 @@ describe('NxSubmitMask', function() {
   it('passes the successMessage prop as the child of NxLoadingSpinner when the success prop is true', function() {
     const spinnerChildren = shallow(<NxSubmitMask message="foo" successMessage="bar" success />)
-        .find(NxLoadingSpinner).children();
+        .find('div.nx-submit-mask__message span').children();
```
Selecting unclassed elements like this isn't terribly maintainable.  I'd rather there be a class added to the span.

glenhunter replied:
> Selecting unclassed elements like this isn't terribly maintainable. I'd rather there be a class added to the span.

Done: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/7eaad8766f8001a37f8fff6f890d10ea7a8be8af

rpokorny commented on lib/src/components/NxSubmitMask/__tests__/NxSubmitMask.test.tsx:
```diff
@@ -11,14 +11,13 @@ import NxLoadingSpinner from '../../NxLoadingSpinner/NxLoadingSpinner';
-        expect(component.find('h3')).toHaveClassName('nx-submit-mask__message');
-        expect(component.find('h3').find(NxLoadingSpinner)).toExist();
+        expect(component.find('div.nx-submit-mask div')).toHaveClassName('nx-submit-mask__message');
+        expect(component.find('div.nx-submit-mask__message').find(NxLoadingSpinner)).toExist();
```
Btw you don't need to add the `div` part at the beginning, you can start with just the classname `.nx-submit-mask__message`.  That's what we'd typically do

glenhunter replied:
> Btw you don't need to add the `div` part at the beginning, you can start with just the classname `.nx-submit-mask__message`. That's what we'd typically do

It didn't work for me before, but now it does so I must have screwed something else up: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/7eaad8766f8001a37f8fff6f890d10ea7a8be8af

rpokorny commented on lib/src/components/NxSubmitMask/__tests__/NxSubmitMask.test.tsx:
```diff
@@ -42,17 +41,18 @@ describe('NxSubmitMask', function() {
 
   it('passes the successMessage prop as the child of NxLoadingSpinner when the success prop is true', function() {
```
This test name is no longer accurate

glenhunter replied:
> This test name is no longer accurate

https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/1b1a50bff29509928cd95a565554b8b8af968045

glenhunter commented:
> The Placement/Layout section of the mockups indicates that the message box should be 200px from the top of the viewport. Fullscreen submit message boxes are currently centered in the viewport. I'm also wondering if @dturner32 is aware that the masks are not fullscreen by default, and whether he intends that we continue to support that or that we drop it.

I adjusted it so that the message is now `200px` from the top in fullscreen mode.

glenhunter commented:
> Success icon has left margin; it shouldn't
> ![image](https://user-images.githubusercontent.com/3630091/97229337-98309080-17ae-11eb-8f6d-09a5e940f9a4.png)

I adjusted this, and also fixed the icon right spacing here and in NxLoadingSpinner since I assume it should be the same, right @dturner32 ? https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/ffc08f8c85b6cebc446c70318c0baaa27faf0011

rpokorny commented on lib/src/components/NxSubmitMask/NxSubmitMask.scss:
```diff
@@ -26,21 +27,29 @@
   position: fixed;
+
+  .nx-submit-mask__message {
+    position: absolute;
```
It seems slightly weird that the positioning of this element is done absolutely in this case, but using flexbox in the normal case.  Could use `align-self` and `margin-top` to keep it as flexbox in both cases.

glenhunter replied:
> It seems slightly weird that the positioning of this element is done absolutely in this case, but using flexbox in the normal case. Could use `align-self` and `margin-top` to keep it as flexbox in both cases.

Flexbox aligning: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/457c922130aba6edad12f7121732a4530f493164

rpokorny commented on lib/src/components/NxSubmitMask/NxSubmitMask.tsx:
```diff
@@ -23,14 +25,19 @@ const NxSubmitMask: FunctionComponent<Props> =
+              <NxFontAwesomeIcon icon={faCheckCircle} className="nx-loading-spinner__icon" />
+              <span className="nx-submit-mask__message-text">{showMessage}</span>
+            </> :
+            <NxLoadingSpinner>{showMessage || 'Submitting...'}</NxLoadingSpinner>
```
```suggestion
            <NxLoadingSpinner>{showMessage || 'Submitting…'}</NxLoadingSpinner>
```

rpokorny commented on lib/src/components/NxSubmitMask/NxSubmitMask.tsx:
```diff
@@ -23,14 +25,19 @@ const NxSubmitMask: FunctionComponent<Props> =
+              <NxFontAwesomeIcon icon={faCheckCircle} className="nx-loading-spinner__icon" />
+              <span className="nx-submit-mask__message-text">{showMessage}</span>
+            </> :
+            <NxLoadingSpinner>{showMessage || 'Submitting...'}</NxLoadingSpinner>
```
So we are changing the default text here?

glenhunter replied:
> So we are changing the default text here?

Yes, I think it's odd for the default task of the Submit Mask to be "Loading...", it isn't loading, it's submitting.

rpokorny commented on lib/src/components/NxSubmitMask/NxSubmitMask.tsx:
```diff
@@ -23,14 +25,19 @@ const NxSubmitMask: FunctionComponent<Props> =
 
-        // the default Loading... message is provided by NxLoadingSpinner
-        spinnerMessage = success ? (successMessage || 'Success!') : message;
+        showMessage = success ? (successMessage || 'Success!') : message;
```
This is still confusing.  There's no reason to conflate the `message` and `successMessage` into the same variable.

glenhunter replied:
> This is still confusing. There's no reason to conflate the `message` and `successMessage` into the same variable.

How is this? https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/d88bd9fe8a996785f7bf86a97feed31447f2250d

rpokorny replied:
> How is this? d88bd9f

Better but the variables names you've chosen seem suggestive of being booleans, when they are actually strings.  You don't really need the variables at all; you can move the `foo || 'bar'` expressions directly into the JSX.

glenhunter replied:
> > How is this? [d88bd9f](https://github.com/sonatype/sonatype-react-shared-components/commit/d88bd9fe8a996785f7bf86a97feed31447f2250d)
> 
> Better but the variables names you've chosen seem suggestive of being booleans, when they are actually strings. You don't really need the variables at all; you can move the `foo || 'bar'` expressions directly into the JSX.

Oh cool! I can: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/a1c4908b760082684f29f30f6f839b54d5f3f397

rpokorny commented on lib/src/components/NxSubmitMask/__tests__/NxSubmitMask.test.tsx:
```diff
@@ -11,14 +11,13 @@ import NxLoadingSpinner from '../../NxLoadingSpinner/NxLoadingSpinner';
-        expect(component.find('h3')).toHaveClassName('nx-h3');
-        expect(component.find('h3')).toHaveClassName('nx-submit-mask__message');
-        expect(component.find('h3').find(NxLoadingSpinner)).toExist();
+        expect(component.find('.nx-submit-mask div')).toHaveClassName('nx-submit-mask__message');
+        expect(component.find('.nx-submit-mask__message').find(NxLoadingSpinner)).toExist();
```
```suggestion
        expect(component.find('.nx-submit-mask  .nx-submit-mask__message').find(NxLoadingSpinner)).toExist();
```

rpokorny commented on lib/src/components/NxSubmitMask/__tests__/NxSubmitMask.test.tsx:
```diff
@@ -35,24 +34,23 @@ describe('NxSubmitMask', function() {
     expect(spinnerChildren).toHaveText('bar');
     expect(spinnerChildren).not.toHaveText('foo');
   });
 
```
The new default text for the `NxLoadingSpinner` ("Submitting") needs a test

glenhunter replied:
> The new default text for the `NxLoadingSpinner` ("Submitting") needs a test

https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/b118888cce5bb67e6a8ce8fe9b33ed3f72b9c55f

rpokorny commented on lib/src/components/NxSubmitMask/__tests__/NxSubmitMask.test.tsx:
```diff
@@ -35,24 +34,23 @@ describe('NxSubmitMask', function() {
-        expect(shallow(<NxSubmitMask success />).find(NxLoadingSpinner).children()).toHaveText('Success!');
-        expect(shallow(<NxSubmitMask successMessage={null} success />).find(NxLoadingSpinner).children())
-            .toHaveText('Success!');
+        expect(shallow(<NxSubmitMask success />).find('.nx-submit-mask__message .nx-submit-mask__message-text')
+            .children()).toHaveText('Success!');
+        expect(shallow(<NxSubmitMask successMessage={null} success />)
+            .find('.nx-submit-mask__message .nx-submit-mask__message-text').children()).toHaveText('Success!');
```
```suggestion
        expect(shallow(<NxSubmitMask success />).find('.nx-submit-mask__message-text')).toHaveText('Success!');
        expect(shallow(<NxSubmitMask successMessage={null} success />)
            .find('.nx-submit-mask__message-text')).toHaveText('Success!');
```

glenhunter commented:
> Should the content be centered when it is less than the minimum width? It isn't currently.

Done here: https://github.com/sonatype/sonatype-react-shared-components/pull/182/commits/5c7e6fd28e6027c5593e578973c6951fc89a9438
